src/plugins/data/public/ui/search_bar_extensions/search_bar_extension.tsx

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { EuiErrorBoundary } from '@elastic/eui';
+import React, { useEffect, useMemo, useRef, useState } from 'react';
+import ReactDOM from 'react-dom';
+import { IIndexPattern } from '../../../common';
+import { DataSource } from '../../data_sources/datasource';
+
+interface SearchBarExtensionProps {
+  config: SearchBarExtensionConfig;
+  dependencies: SearchBarExtensionDependencies;
+  componentContainer: Element;
+  bannerContainer: Element;
+}
+
+export interface SearchBarExtensionDependencies {
+  /**
+   * Currently selected index patterns.
+   */
+  indexPatterns?: Array<IIndexPattern | string>;
+  /**
+   * Currently selected data source.
+   */
+  dataSource?: DataSource;
+  /**
+   * Currently selected query language.
+   */
+  language: string;
+}
+
+export interface SearchBarExtensionConfig {
+  /**
+   * The id for the search bar extension.
+   */
+  id: string;
+  /**
+   * Lower order indicates higher position on UI.
+   */
+  order: number;
+  /**
+   * A function that determines if the search bar extension is enabled and should be rendered on UI.
+   * @returns whether the extension is enabled.
+   */
+  isEnabled: (dependencies: SearchBarExtensionDependencies) => Promise<boolean>;
```

Comment:
nit: should we consider making this a behavior subject? then appending $ like `isEnabled$` might be easier to distinguish that i should subscribe to it.

Reply:
i haven't used observables much, a quick search tells me to use `const isEnabled$ = (dependencies) => from(isEnabled(dependencies));` which i can't see what the benefit is. could you elaborate more a bit?